src/main/java/backend/Logic.java

```diff
@@ -200,18 +198,51 @@ public Model getRepo(String repoId) {
      * @return true if label replacement on GitHub was a success, false otherwise.
      */
     public CompletableFuture<Boolean> replaceIssueLabels(TurboIssue issue, List<String> newLabels) {
+        List<String> originalLabels = issue.getLabels();
+
         logger.info("Changing labels for " + issue + " on UI");
-        issue.setLabels(newLabels);
+        Optional<TurboIssue> replaceResult = models.replaceIssueLabels(issue.getRepoId(), issue.getId(), newLabels);
+        if (!replaceResult.isPresent()) {
+            return CompletableFuture.completedFuture(false);
+        }
         refreshUI();
 
-        return updateIssueLabelsOnRepo(issue, newLabels);
+        return updateIssueLabelsOnRepo(issue, newLabels)
+                .thenApply((isUpdateSuccessful) -> {
+                    if (isUpdateSuccessful) {
+                        return true;
+                    }
+                    logger.error("Unable to update model on server");
+                    revertLocalLabelsReplace(replaceResult.get(), originalLabels);
+                    return false;
+                });
+    }
+
+    private Optional<TurboIssue> lookUpIssue(String repoId, int issueId) {
+        Optional<Model> modelLookUpResult = models.getModelById(repoId);
+        if (!modelLookUpResult.isPresent()) {
+            logger.error("Model " + repoId + " not found in models");
```

Comment:
Maybe we could factor out this pattern of mapping over the optional, but logging if it's empty? Could be a utility method, since I think Optional::map just ignores the empty case.